## Implementing a routine to compute vector length

### Preliminaries

<p><font color=red> Again, copy this notebook so that you don't corrupt the original!  Then you can "play" with the copy of the notebook all you want! </font> </p>

<p>
<b>
NOTE: A common problem that students have with IPython notebooks is not understanding that when the code in the gray boxes (cells) is executed, it assigns variables that persist the whole time that the notebook is open. Further, some cells rely on variables assigned by earlier cells.  If you execute these cells out of order, or if you execute the same cell twice, then you may end up changing the value of the variables.  To correct this, click on "Cell" at the top and execute "run all above" or "run all".  You can also reset all cells by clicking "Cell -> All Output -> Clear"
</b>
</p>

<p>In this notebook, we show how to write a routine to compute vector length. </p>

Let's start by importing numpy and creating a vector $ x = \left( \begin{array}{r} 1 \\ 2 \\ 3 \end{array} \right) $.

Execute the code in the box by clicking in the box and then on "Cell -> Run".  Alternative, click on the box and push "Shift" and "Return" (or "Enter") together.

In [1]:
import numpy as np    # This imports a package called "numpy" that will make working with matrices 
                      # simpler

# create a two-dimensional matrix with only one column.  
x = np.matrix( '1.;2.;3.' )
print( 'x = ' )
print( x )

x = 
[[1.]
 [2.]
 [3.]]


Now, recall that the length of a vector equals $ \| x \|_2 = \sqrt{ x^T x } $.  So, we can use the dot product routine that we wrote before to compute the length:

In [2]:
import laff
import math

length_x = math.sqrt( laff.dot( x, x ) )

print( 'length_x:' )
print( length_x )

length_x:
3.7416573867739413


### Length as a simple routine

<p>
As before, it is preferable to create a routine that computes the length.
</p>

<p>
Complete the following routine to implement this:
</p>

In [3]:
def length( x ):
    #Your code here
    alpha = 0
    for i in range(len(x)):
        alpha += x[i, 0] ** 2
    res = np.sqrt(alpha)
    return res

Be sure the run the above box, or this notebook won't know about the routine!!!

Now, if you execute

In [4]:
length_x = length( x )

print( 'length_x:' )
print( length_x )

print( 'difference between length_x and math.sqrt( laff.dot( x, x ) ):' )
print( length_x - math.sqrt( laff.dot( x, x ) ) )

length_x:
3.7416573867739413
difference between length_x and math.sqrt( laff.dot( x, x ) ):
0.0


The result should be:

<code>
length_x:
3.7416573867739413
difference between length_x and math.sqrt( laff.dot( x, x ) ):
0.0
</code>

### A complete length function as part of the LAFF library

As we proceed with progressively more advanced operations and routines, we are going to need a general dot routine where $ x $ and $ y $ can be row and/or column vectors.  

This routine is part of the 'laff' library.  If you do

<code>
import laff
</code>

then <code> laff.norm2( x ) </code> will perform the desired computation of the length of $ x $, when <code> x </code> is a column and/or a row vectors.  If you really want to see what this routine looks like, then ask for it on the discussion forum and we'll point you to where it can be found.

In [5]:
import laff

length_x = laff.norm2( x )

print( 'length_x:' )
print( length_x )

print( 'difference between length_x and math.sqrt( laff.dot( x, x ) ):' )
print( length_x - math.sqrt( laff.dot( x, x ) ) )

length_x:
3.741657386773941
difference between length_x and math.sqrt( laff.dot( x, x ) ):
-4.440892098500626e-16


### Need a challenge?

In "1.7.3 Overflow and Underflow", we discuss how computing the length with the dot product can inadvertently cause overflow or underflow.  Write a routine that avoids that

In [ ]:
def length( x ):
    ### You fill in the rest!